### PR TITLE
feat: add slurm cluster autodiscovery

### LINKED ISSUE #7: Implement SLURM cluster autodiscovery
Query a local SLURM scheduler and return a structured representation of available resources (partitions, node types, accounts, QOS policies, GPU types).

## Motivation

Every SLURM-facing feature requires the user to manually specify partition names, account strings, GPU types, and core
counts. SLURM already exposes this via `sinfo` and `sacctmgr`. Autodiscovery removes redundant manual configuration and
 enables downstream features to make resource-aware decisions automatically.

**No dependencies.** This issue also creates the `mdfactory/performance/` package.

## Scope

- Create `mdfactory/performance/__init__.py`
- `mdfactory/performance/cluster.py`
- Dataclasses: `NodeType`, `Partition`, `ClusterInfo`
- `discover_cluster() -> ClusterInfo` — parses sinfo/sacctmgr output
- `select_partition(cluster, needs_gpu, min_cpus, min_mem_gb) -> Partition | None`
- Graceful degradation: returns None when SLURM commands unavailable
- Session-level caching (cluster topology doesn't change mid-session)

## SLURM commands used

| Command | Information extracted |
|---------|---------------------|
| `sinfo -N --noheader -o "%P %n %c %m %G %f %l %T"` | Partition, node, CPUs, memory, GPUs (gres), features, timelimit,
 state |
| `sacctmgr show assoc user=$USER format=Account --noheader --parsable2` | Accounts available to current user |
| `sacctmgr show qos format=Name,MaxWall,MaxTRES --noheader --parsable2` | QOS policies with limits |

## Test structure

Tests use flat files: `mdfactory/tests/test_cluster.py` (not a subdirectory).

## Acceptance criteria

- `from mdfactory.performance import cluster` imports without error
- Unit tests with mocked sinfo/sacctmgr output pass
- Returns None gracefully on non-SLURM machine (laptop)
- Parses real sinfo output format correctly (multi-GPU nodes, mixed partitions)

### PR DESCRIPTION
**Pull request by gregorweiss**
Adds `mdfactory/performance/cluster.py` — queries `sinfo` and `sacctmgr` to discover cluster resources automatically. This is the foundation for all planned HPC features (benchmarking, node packing, GPU sharing).

## What it does

- `discover_cluster()` → returns partitions, node specs (CPUs, memory, GPUs), accounts, and QOS policies as frozen dataclasses
- `select_partition(cluster, needs_gpu=True, min_cpus=64)` → picks the best partition for given requirements
- Returns `None` gracefully on non-SLURM machines
- Results cached per session (topology doesn't change mid-run)

## Files

- `mdfactory/performance/__init__.py` — new package
- `mdfactory/performance/cluster.py` — autodiscovery implementation
- `mdfactory/tests/test_cluster.py` — unit tests with mocked SLURM output

## Test plan

- [x] `pytest mdfactory/tests/test_cluster.py -v` (mocked, no cluster needed)
- [x] Verify `discover_cluster()` returns `None` on laptop
- [x] Verify populated `ClusterInfo` on a real SLURM node

### CODE REVIEW

File: mdfactory/performance/cluster.py [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+        SLURM feature/constraint tags on this node type (immutable).
+    count : int
+        Number of nodes with this exact configuration.
+    """
+
+    cpus: int
+    memory_mb: int
+    gpu_specs: tuple[tuple[int, str], ...] = field(default_factory=tuple)
+    features: tuple[str, ...] = field(default_factory=tuple)
+    count: int = 1
```
Comment (maxscheurer) [inline mdfactory/performance/cluster.py]: We use `pydantic` everywhere else in `mdfactory`, so would be great to use that throughout.

File: mdfactory/performance/cluster.py [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+        The user's default account, if determinable.
+    """
+
+    partitions: list[Partition] = field(default_factory=list)
+    accounts: list[str] = field(default_factory=list)
+    qos_policies: list[str] = field(default_factory=list)
+    default_account: str | None = None
+
+
+def _run_command(cmd: list[str], *, timeout: int = 30) -> str | None:
```
Comment (maxscheurer) [inline mdfactory/performance/cluster.py]: Check if we have this utility function somewhere else already.

Comment (gregorweiss) [reply mdfactory/performance/cluster.py]: Good catch! I've refactored this into a generalized `run_command()` utility in `mdfactory/utils/utilities.py`. 

The cluster autodiscovery now uses this instead of its local `_run_command()`. I also identified several other subprocess calls that could benefit from this utility (in `cli.py`, `sync_config.py`, and `bilayer/artifacts/__init__.py`), but didn't want to expand the scope of this PR beyond cluster autodiscovery.

I've created issue #19 to track migrating those locations separately - keeping this PR focused while ensuring the broader refactoring doesn't get lost.

File: mdfactory/performance/cluster.py (hunk 1 above)
Comment (gregorweiss) [reply mdfactory/performance/cluster.py]: Done.

**Comment by gregorweiss** [comment]
The PR #11 scope from emdgroup/mdfactory#20 is complete on this branch: `BaseSlurmConfig` extracted to `performance/slurm_config.py`, `SlurmConfig` migrated to Pydantic, `normalize_slurm_time` moved, `from_yaml()` added, `config.ini` split into `PARTITION_CPU`/`PARTITION_GPU`/`DEFAULT_QOS`, tests updated. Re-export shim in `analysis/submit.py` keeps existing callers intact.

**Review by maxscheurer — COMMENTED** [review]
## Code Review

Reviewed by mach6 (multi-agent: code-reviewer, error-auditor, test-reviewer, completeness-checker, simplifier + independent assessment). Overall a solid, well-structured PR — all of issue #20's requirements are met and the Pydantic migration / three-tier `from_cluster` precedence are clean. The inline comments below are the genuine items worth addressing before merge (7 total: 3 robustness, 4 test gaps). Nitpicks and one false positive were filtered out during assessment.

Note: the PR description only mentions `cluster.py`, but the change is much broader (full BaseSlurmConfig unification + a new user-facing `config cluster` CLI command + behavioral changes to `analysis run`/`artifacts run` where `--account` is now optional). Worth expanding the description so reviewers know the full surface.

---
*Reviewed by mach6*

File: mdfactory/performance/cluster.py [hunk 3]
```diff
@@ -0,0 +1,651 @@
...
+        timeout=30,
+        graceful=True,
+    )
+    if output is None:
+        return None
+    account = output.strip().splitlines()[0].strip() if output.strip() else None
+    return account if account else None
+
+
+@functools.lru_cache(maxsize=1)
```
Comment (maxscheurer) [inline mdfactory/performance/cluster.py]: **Robustness (medium):** `@lru_cache(maxsize=1)` caches *any* return value, including `None`. There are two very different `None` paths: (a) `sinfo` genuinely absent (line 564 — intentional, fine to cache), and (b) `sinfo` present but `_discover_partitions()` returned `None` because `sinfo` exited non-zero (line 568 — a *transient* failure, e.g. `slurmctld` restart or a momentary NFS/permission hiccup).

The transient `None` gets cached for the whole process. Because `from_cluster()` is called per-job during batch submission within a single process, one momentary blip makes every subsequent resolution fail with "SLURM autodiscovery failed and no account configured" even after the cluster recovers. `cache_clear()` exists but isn't called on retry.

Suggest caching only successful `ClusterInfo` results — e.g. move the `shutil.which` guard to an outer un-cached wrapper and only apply `lru_cache` to the success path, or don't cache the transient-failure branch.

File: mdfactory/performance/cluster.py [hunk 4]
```diff
@@ -0,0 +1,651 @@
...
+        if is_default:
+            partition_data[partition_name]["is_default"] = True
+
+        # Determine if node is operational (can schedule jobs or report accurate config)
+        # Exclude only truly broken/invalid nodes: down, drained, inval, fail, maint, unknown
+        state_lower = state.lower()
+        is_broken = any(
+            broken in state_lower
+            for broken in ("down", "drained", "inval", "fail", "maint", "unknown")
+        )
```
Comment (maxscheurer) [inline mdfactory/performance/cluster.py:331]: **Robustness (medium):** Bare substring matching misclassifies power-management states. On SLURM ≥ 21.08 with power saving / cloud bursting, `%T` reports `powered_down` and `powering_down` — both contain `"down"`, so `"down" in state_lower` is `True` and these perfectly schedulable nodes are marked broken. Their specs are then dropped from `node_types`, and a partition whose nodes are all powered down ends up with no operational node → `select_partition` skips it.

The substring approach is deliberate for `fail`→`failing` / `inval`→`invalid`, so a full rewrite isn't needed — just exclude the power states, e.g.:
```python
_POWER_STATES = {"powered_down", "powering_down", "powering_up"}
is_broken = state_lower not in _POWER_STATES and any(
    broken in state_lower for broken in ("down", "drained", "inval", "fail", "maint", "unknown")
)
```
Please add a fixture covering `powered_down` / `powering_down`.

File: mdfactory/performance/cluster.py [hunk 5]
```diff
@@ -0,0 +1,651 @@
...
+
+        # Strip socket binding suffix like "(S:0-1)" or "(IDX:0-3)"
+        if "(" in entry:
+            entry = entry.split("(")[0]
+
+        parts = entry.split(":")
+        if len(parts) == 3:
+            # gpu:type:count
+            _, gpu_type, count_str = parts
+            gpu_entries.append((int(count_str), gpu_type))
```
Comment (maxscheurer) [inline mdfactory/performance/cluster.py]: **Robustness (low-med):** This `int(count_str)` in the 3-part `gpu:type:count` branch is unguarded, unlike the 2-part branch just below (lines 165-170) which wraps `int()` in `try/except ValueError`. A malformed 3-part GRES token raises `ValueError` that propagates out of `_parse_sinfo` → `_discover_partitions` → `discover_cluster` (uncaught), and the CLI only catches `RuntimeError`, so the user gets a raw traceback and *all* partition discovery dies. Suggest applying the same `try/except ValueError` (skip the entry / log a warning) for symmetry.

File: mdfactory/performance/cluster.py [hunk 6]
```diff
@@ -0,0 +1,651 @@
...
+        Memory field from sinfo (numeric, in MB by default).
+
+    Returns
+    -------
+    int
+        Memory in MB. Returns 0 on parse failure.
+    """
+    try:
+        # sinfo %m gives memory in MB as an integer
+        cleaned = mem_str.strip().rstrip("+")
```
Comment (maxscheurer) [inline mdfactory/performance/cluster.py:217]: **Test gap (low):** The trailing-`+` stripping handles SLURM's `128000+` notation (memory varies across nodes in a partition), but no fixture exercises it. One-line test: assert `_parse_memory_mb("128000+") == 128000`.

File: mdfactory/cli.py [hunk 7]
```diff
@@ -1233,16 +1231,40 @@ def analysis_run(
...
+                f"partition={slurm_cfg.partition}"
+            )
+        except RuntimeError as e:
+            raise ValueError(
+                f"SLURM autodiscovery failed: {e}\nPlease specify --account explicitly."
+            ) from e
+    else:
+        slurm_cfg = SlurmConfig(
+            account=account,
+            partition=partition or "cpu",
```
Comment (maxscheurer) [inline mdfactory/cli.py:1260]: **Test gap (medium):** This autodiscovery glue is untested — `test_cli_cluster.py` only covers `config cluster`. The underlying `from_cluster()` is tested, but not this wrapper: the `account is None` branch, the `RuntimeError → ValueError` re-wrap with the "Please specify --account explicitly" message, and the `partition or "cpu"` fallback in the explicit-account branch. The identical block in `analysis_artifacts_run` (lines 1545-1571) is also untested. Please add CLI tests for both: autodiscovery success, the rewrap failure path, and partition forwarding.

File: mdfactory/cli.py [hunk 8]
```diff
@@ -1750,6 +1794,112 @@ def config_edit():
...
+        print(f"    Max Time: {partition.max_time}")
+        if partition.default_time != partition.max_time:
+            print(f"    Default Time: {partition.default_time}")
+
+        # Summarize node types
+        for nt in partition.node_types:
+            gpu_info = ""
+            if nt.gpu_specs:
+                # Format multiple GPU types like "7x b200, 7x 1g.23gb"
+                gpu_parts = [f"{count}x {gtype}" for count, gtype in nt.gpu_specs]
```
Comment (maxscheurer) [inline mdfactory/cli.py]: **Display (low):** `gtype` can legitimately be `None` (`_parse_gres` yields `(count, None)` for a typeless `gpu:4` GRES), so this renders as `"4x None"`. Use `f"{count}x {gtype or 'unknown'}"`.

File: mdfactory/performance/slurm_config.py [hunk 9]
```diff
@@ -0,0 +1,339 @@
...
+        resolved_partition: str | None = extra_fields.pop("partition", None)
+        if resolved_partition is None:
+            resolved_partition = (
+                settings.slurm_partition_gpu if needs_gpu else settings.slurm_partition_cpu
+            )
+        if resolved_partition is None:
+            if cluster is None:
+                raise RuntimeError(
+                    "SLURM autodiscovery failed and no partition configured. "
+                    "Set [slurm] PARTITION_CPU or PARTITION_GPU in config.ini "
```
Comment (maxscheurer) [inline mdfactory/performance/slurm_config.py:206]: **Test gap (low-med):** This "SLURM autodiscovery failed and no partition configured" path (account resolves from config, partition unset, `cluster is None`) is never hit by tests — the existing no-suitable-partition test uses a *present* cluster and exercises the `select_partition → None` branch instead (line 217). This is a real HPC scenario (controller down but account configured manually). Add a test: `discover_cluster` mocked to `None`, `slurm_account` set, `slurm_partition_cpu` unset → assert this RuntimeError.

File: mdfactory/tests/test_cluster.py [hunk 10]
```diff
@@ -0,0 +1,526 @@
...
+gpu node012 96 512000 gpu:h100:8 h100,nvlink 1-00:00:00 4:00:00 mixed
+bigmem node020 256 2048000 (null) bigmem,epyc 7-00:00:00 1-00:00:00 idle
+"""
+
+SINFO_OUTPUT_SINGLE_PARTITION = """\
+compute node001 64 128000 (null) (null) 2-00:00:00 1-00:00:00 idle
+compute node002 64 128000 (null) (null) 2-00:00:00 1-00:00:00 idle
+"""
+
+SINFO_OUTPUT_GPU_ONLY = """\
```
Comment (maxscheurer) [inline mdfactory/tests/test_cluster.py:44]: **Test gap (low):** `SINFO_OUTPUT_GPU_ONLY` is defined but never referenced — it's dead code. It's the ideal fixture for exercising `select_partition`'s node-count tie-break (`gpu-short` 2 nodes vs `gpu-long` 1 node, both non-default), which the sort key `(not is_default, -total_nodes, name)` handles but no test covers. Add a test asserting `gpu-short` is chosen for `needs_gpu=True`.

File: mdfactory/performance/cluster.py (hunk 4 above)
Comment (gregorweiss) [reply mdfactory/performance/cluster.py:331]: Intentional: `powered_down` and `powering_down` are correctly caught by the `"down"` substring check and should remain excluded. Wakeup time is unpredictable — SLURM power-saving wakes nodes in minutes, but an admin-initiated power-down (same state string) can mean days. The substring match encodes the right conservative semantics without needing to enumerate power-management states explicitly. Adding `"error"` to the list is the only genuine gap here.

File: mdfactory/performance/cluster.py (hunk 3 above)
Comment (gregorweiss) [reply mdfactory/performance/cluster.py]: Fixed: replaced `@lru_cache` with a mutable list cell (`_cluster_cache`). Only confirmed sinfo-absent and a successful `ClusterInfo` are cached; transient sinfo failures fall through uncached and retry on next call. `cache_clear()` attached as a function attribute to preserve the existing test API.

File: mdfactory/performance/cluster.py (hunk 5 above)
Comment (gregorweiss) [reply mdfactory/performance/cluster.py]: Fixed: wrapped in `try/except ValueError: continue`, matching the 2-part branch directly below.

File: mdfactory/performance/cluster.py (hunk 6 above)
Comment (gregorweiss) [reply mdfactory/performance/cluster.py:217]: Fixed: added `test_memory_trailing_plus_stripped` in `TestParseSinfo`.

File: mdfactory/cli.py (hunk 7 above)
Comment (gregorweiss) [reply mdfactory/cli.py:1260]: Fixed: added `TestAnalysisRunSlurmGlue` and `TestAnalysisArtifactsRunSlurmGlue` in `test_cli_cluster.py`, covering the autodiscovery path, `RuntimeError → ValueError` re-raise, and explicit-account path for both commands.

File: mdfactory/cli.py (hunk 8 above)
Comment (gregorweiss) [reply mdfactory/cli.py]: Fixed: `f"{count}x {gtype or 'unknown'}"`.

File: mdfactory/performance/slurm_config.py (hunk 9 above)
Comment (gregorweiss) [reply mdfactory/performance/slurm_config.py:206]: Fixed: added `test_autodiscovery_fails_no_partition_in_config_raises` in `TestBaseSlurmConfigFromCluster` — patches `discover_cluster → None` and supplies account via `Settings` so the failure reaches the partition step.

File: mdfactory/tests/test_cluster.py (hunk 10 above)
Comment (gregorweiss) [reply mdfactory/tests/test_cluster.py:44]: Fixed: used in new `test_node_count_tiebreak` in `TestSelectPartition` — asserts `gpu-short` (2 nodes) is preferred over `gpu-long` (1 node) when both qualify.

**Review by maxscheurer — APPROVED** [review]

**Comment by maxscheurer** [comment]
@gregorweiss, please add `develop` to the list of pull request branches here, so the CI runs:

https://github.com/emdgroup/mdfactory/blob/b59eccb681e00e78dcc118a5e6f1f07e832975f7/.github/workflows/ci.yml#L11
